encoding.go

```diff
@@ -106,9 +107,17 @@ func encodeUvarintSize(u uint64) int {
 	return (bits.Len64(u) + 6) / 7
 }
 
+var encodeVarintPool = &sync.Pool{
+	New: func() interface{} {
+		return &[binary.MaxVarintLen64]byte{}
+	},
+}
+
 // encodeVarint writes a varint-encoded integer to an io.Writer.
 func encodeVarint(w io.Writer, i int64) error {
-	var buf [binary.MaxVarintLen64]byte
+	buf := encodeVarintPool.Get().(*[binary.MaxVarintLen64]byte)
+	defer encodeVarintPool.Put(buf)
```

Comment:
is there a real difference here? We are talking about allocating 10 bytes on on the stack. I think it's faster then going through a sync pool.

Reply:
@robert-zaremba Yes, if you look at the issue for profile mem:

```
13.21GB  4.79% 52.74%    24.46GB  8.87%  github.com/cosmos/iavl.encodeVarint
```

Reduce 30% of 24GB will be a huge win.

What we deal is actually the number of allocations, since when we nearly don't have concurrently access to the pool, so the speed will be nearly identical.

If someday, we need to support concurrently call to encodeVarint, then we can change it to accept a `*[binary.MaxVarintLen64]` as input, so the caller can allocate just one and pass the same pointer each call to encodeVarint.

Reply:
Good tip. Could you add a comment about this in the code, please?

Reply:
Done